app/pages/learn/01_tutorial/08_more-resources/02_debugging.md

```diff
@@ -0,0 +1,136 @@
+slug: learn/debugging
+type: tutorial
+category: resources
+layout: learn/tutorial.html
+subheader_select: tutorials
+main_css_id: learn
+toc:
+- What is debugging? {intro}
+- Why not println {println}
+- Why not unit testing {unit}
+- What is a breakpoint? {breakpoint}
+- How can debuggers be used? {usage}
+- Debugger basics {basic}
+- Advanced techniques {advanced}
+- Documentation {docs}
+description: "Learning how to use a debugger"
+last_update: 2023-11-05
+author: ["JeanneBoyarsky"]
+---
+
+<a id="intro">&nbsp;</a>
+## What is debugging?
+
+We all write perfect code that works on the first attempt, right? Ha! Just kidding. We often have to find and fix errors in our code. This process is called debugging. 
+
+You might be wondering why it is called "debugging". The term became popular in the 1940s after Admiral Grace Hopper found a moth inside a computer (remember computers were giant back then so an actual insect could get in.) While you won't have to deal with animals in your code, we will have to deal with errors and problems that we refer to as bugs. You may also hear them called defects although that's usually after you've committed the code. Regardless of the name of the issue, you still have to find that problem and that's debugging!
+
+A debugger is a tool in your IDE (integrated development environment) that lets you see the values of different variables at different points in the program. It's like a really powerful magnifying glass. While a few details vary between Eclipse/IntelliJ/NetBeans/VS Code, the concepts are the same.
+
+<a id="println">&nbsp;</a>
+## Why not Println
+
+When first learning how to code, we often write code like this to see what is going on:
+
+```java
+System.out.println(numCats);
+```
+
+There's nothing wrong with using println (as long as you don't commit it). However, it can quickly get unmanagable in a more complicated program especally if you have multiple things to keep track of or a lot of loops. It can also be hard to find when there is a lot of logging in the application. (I used the stars and my initials to mitigate that, but still can get past the point where println is useful.)
+
+```java
+System.out.println("*** JB i=%s numCats=%s numDogs=%s".formatted(i, numCats, numDogs));
+```
+
+Even if println() does meet your needs at the moment, it won't forever. Learning how to use a debugger helps avoid the problem that you use println just because it is all you know.
+
+<a id="unit">&nbsp;</a>
+## Why not unit testing
+
+Nothing is wrong with unit testing. Unit tests are great. They document expected behavior. They tell you if an unexpected value is returned. They help you understand the behavior of the code. And sometimes they can even give you big clues about what is wrong with the code. However, they don't tell you what is happening inside the broken code when it isn't returning the right value. For that, you use a debugger with the unit test to see what is going on inside the method.
```

Comment:
Maybe also add a sentence of: "And once you fix your code, unit tests can help catch new bugs from being introduced."?